### PR TITLE
Added `oryx exec` command

### PR DESCRIPTION
**Pull request by dorfire**

### CODE REVIEW

File: src/BuildScriptGeneratorCli/Commands/ExecCommand.cs [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+using Microsoft.Extensions.Options;
+using Microsoft.Oryx.BuildScriptGenerator;
+using Microsoft.Oryx.Common;
+using Microsoft.Oryx.Common.Extensions;
+
+namespace Microsoft.Oryx.BuildScriptGeneratorCli
+{
+    [Command(Name, Description = "Execute an arbitrary command in the default shell, in an environment " +
+        "with the best-matching platform tool versions.")]
+    internal class ExecCommand : CommandBase
```
Comment (kichalla) [inline src/BuildScriptGeneratorCli/Commands/ExecCommand.cs]: Tests? There are fair number of cases we need to test here. commands with quotes, multiple commands separated by `&&`, a script file itself

File: src/BuildScriptGeneratorCli/Commands/ExecCommand.cs [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+
+        [Argument(0, Description = "The source directory.")]
+        public string SourceDir { get; set; }
+
+        [Argument(1, Description = "The command to execute in an app-specific environment.")]
+        public string Command { get; set; }
+
+        internal override int Execute(IServiceProvider serviceProvider, IConsole console)
+        {
+            var logger = serviceProvider.GetRequiredService<ILogger<BuildCommand>>();
```
Comment (kichalla) [inline src/BuildScriptGeneratorCli/Commands/ExecCommand.cs]: `ILogger<ExecCommand>`

File: src/BuildScriptGenerator/CheckerMessage.cs [hunk 3]
```diff
@@ -4,24 +4,19 @@
 // --------------------------------------------------------------------------------------------
 
 using Microsoft.Extensions.Logging;
-using Microsoft.Oryx.BuildScriptGenerator;
 
 namespace Microsoft.Oryx.BuildScriptGenerator
 {
     public class CheckerMessage : ICheckerMessage
     {
-        private readonly LogLevel _level;
```
Comment (kichalla) [inline src/BuildScriptGenerator/CheckerMessage.cs]: As we discussed earlier in my PR, let's avoid making unrelated changes. A separate clean-up PR would be better for these.

File: src/BuildScriptGenerator/DefaultBuildScriptGenerator.cs [hunk 4]
```diff
@@ -241,8 +241,8 @@ private static string GetBenvArgs(IDictionary<string, string> benvArgsMap)
         private IList<BuildScriptSnippet> GetBuildSnippets(
             BuildScriptGeneratorContext context,
             Dictionary<string, string> toolsToVersion,
-            List<string> directoriesToExcludeFromCopyToIntermediateDir,
-            List<string> directoriesToExlcudeFromCopyToBuildOutputDir)
+            [CanBeNull] List<string> directoriesToExcludeFromCopyToIntermediateDir,
```
Comment (kichalla) [inline src/BuildScriptGenerator/DefaultBuildScriptGenerator.cs]: Please revert all these type of unrelated changes!

File: src/BuildScriptGenerator/SourceRepo/DefaultSourceRepoProvider.cs [hunk 5]
```diff
@@ -39,42 +31,5 @@ public ISourceRepo GetSourceRepo()
             _sourceRepo = new LocalSourceRepo(_sourceDirectory, _loggerFactory);
             return _sourceRepo;
         }
-
-        private void CopyDirectories(string sourceDirectory, string destinationDirectory, bool recursive)
```
Comment (kichalla) [inline src/BuildScriptGenerator/SourceRepo/DefaultSourceRepoProvider.cs]: Was this dead code?

File: src/BuildScriptGeneratorCli/Commands/CommandBase.cs [hunk 6]
```diff
@@ -24,7 +24,7 @@ internal abstract class CommandBase
         public string LogFilePath { get; set; }
 
         [Option("--debug", Description = "Print stack traces for exceptions.")]
```
Comment (kichalla) [inline src/BuildScriptGeneratorCli/Commands/CommandBase.cs]: I am not suggesting to make a change here, but I think this should probably be `--debugInfo`

File: src/BuildScriptGeneratorCli/Commands/ExecCommand.cs (hunk 2 above)
Comment (dorfire) [reply src/BuildScriptGeneratorCli/Commands/ExecCommand.cs]: yeah, fixed already but did not push. thanks!

File: src/BuildScriptGenerator/DefaultBuildScriptGenerator.cs (hunk 4 above)
Comment (dorfire) [reply src/BuildScriptGenerator/DefaultBuildScriptGenerator.cs]: They aren't unrelated - i changed the way this function works, so that it can accept `null`s in these two parameters.
The new method `GetRequiredToolVersions` relies on this change.

File: src/BuildScriptGeneratorCli/Commands/ExecCommand.cs (hunk 1 above)
Comment (dorfire) [reply src/BuildScriptGeneratorCli/Commands/ExecCommand.cs]: working on them still; didn't want to block the review of this PR in the meantime

File: src/BuildScriptGenerator/SourceRepo/DefaultSourceRepoProvider.cs (hunk 5 above)
Comment (dorfire) [reply src/BuildScriptGenerator/SourceRepo/DefaultSourceRepoProvider.cs]: yep

File: src/BuildScriptGenerator/DefaultBuildScriptGenerator.cs [hunk 7]
```diff
@@ -241,8 +241,8 @@ private static string GetBenvArgs(IDictionary<string, string> benvArgsMap)
...
+            [CanBeNull] List<string> directoriesToExlcudeFromCopyToBuildOutputDir)
```
Comment (gpcastro) [inline src/BuildScriptGenerator/DefaultBuildScriptGenerator.cs]: can you also fix the typo: directoriesToEx*cl*ude

File: src/BuildScriptGeneratorCli/Commands/ExecCommand.cs [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+{
+    [Command(Name, Description = "Execute an arbitrary command in the default shell, in an environment " +
+        "with the best-matching platform tool versions.")]
+    internal class ExecCommand : CommandBase
+    {
+        public const string Name = "exec";
+
+        public const string DefaultBashPath = "/bin/bash";
+
+        public const string SrcDirDoesNotExistErrorMessageFmt = "Could not find the source directory '{0}'.";
```
Comment (gpcastro) [inline src/BuildScriptGeneratorCli/Commands/ExecCommand.cs]: move those to the resource file with the labels

File: src/BuildScriptGeneratorCli/Commands/ExecCommand.cs [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+        public const string CommandMissingErrorMessage = "A command is required.";
+        public const string NoToolsDetectedErrorMessage = "No usable tools detected for source directory.";
+
+        [Argument(0, Description = "The source directory.")]
+        public string SourceDir { get; set; }
```
Comment (gpcastro) [inline src/BuildScriptGeneratorCli/Commands/ExecCommand.cs]: I would have this is a named parameter, not as positional argument. e.g. `oryx exec -sourceDir /app <command>`

Comment (gpcastro) [reply src/BuildScriptGeneratorCli/Commands/ExecCommand.cs]: and default to current directory

File: src/BuildScriptGeneratorCli/Commands/ExecCommand.cs [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+
+            return exitCode;
+        }
+
+        internal override bool IsValidInput(IServiceProvider serviceProvider, IConsole console)
+        {
+            var logger = serviceProvider.GetRequiredService<ILogger<ExecCommand>>();
+            var options = serviceProvider.GetRequiredService<IOptions<BuildScriptGeneratorOptions>>().Value;
+
+            if (!Directory.Exists(options.SourceDir))
```
Comment (gpcastro) [inline src/BuildScriptGeneratorCli/Commands/ExecCommand.cs]: how about the [DirectoryExists] attribute?

File: src/BuildScriptGeneratorCli/Commands/ExecCommand.cs [hunk 11]
```diff
@@ -0,0 +1,113 @@
...
+            {
+                logger.LogError("Could not find the source directory {srcDir}", SourceDir);
+                console.WriteErrorLine(string.Format(SrcDirDoesNotExistErrorMessageFmt, SourceDir));
+                return false;
+            }
+
+            if (string.IsNullOrWhiteSpace(Command))
```
Comment (gpcastro) [inline src/BuildScriptGeneratorCli/Commands/ExecCommand.cs]: I wouldn't error out in this case, just do a no-op. When writing a script that calls this command, it's easier to just no-op when there is no command so the script doesn't fail. I don't think `benv` for instance fails with an empty command

File: src/BuildScriptGeneratorCli/Commands/ExecCommand.cs (hunk 9 above)
Comment (dorfire) [reply src/BuildScriptGeneratorCli/Commands/ExecCommand.cs]: I wanted to keep this in-line with `BuildCommand`. WDYT?

File: src/BuildScriptGeneratorCli/Commands/ExecCommand.cs (hunk 1 above)
Comment (kichalla) [reply src/BuildScriptGeneratorCli/Commands/ExecCommand.cs]: I see you added only 1 E2E test. Can you add more scenarios as mentioned above?

File: src/BuildScriptGeneratorCli/Commands/ExecCommand.cs [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+                logger.LogInformation("Running {shell} -c {cmd}", shellPath, cmd);
+                if (DebugMode)
+                {
+                    console.WriteLine($"> {shellPath} -c '{cmd}'");
+                }
+
+                exitCode = serviceProvider.GetRequiredService<IScriptExecutor>().ExecuteScript(
+                    shellPath,
+                    new[] { "-c", cmd },
+                    SourceDir,
```
Comment (kichalla) [inline src/BuildScriptGeneratorCli/Commands/ExecCommand.cs]: Doesn't this fail if the SourceDir is not provided by the user? I think you should pick it out of the `options` which resolves the source directory to the current directory if user does not provide it.

File: tests/Oryx.BuildImage.Tests/ExecCommandTest.cs [hunk 13]
```diff
@@ -0,0 +1,57 @@
...
+                .ToString();
+
+            // Act
+            var result = _dockerCli.Run(Settings.BuildImageName, "/bin/bash", "-c", script);
+
+            // Assert
+            RunAsserts(
+                () =>
+                {
+                    Assert.Contains($"{expectedBashPath} -c '{FilePaths.Benv} node={NodeConstants.NodeLtsVersion} {cmd}'", result.StdOut);
```
Comment (kichalla) [inline tests/Oryx.BuildImage.Tests/ExecCommandTest.cs]: nit: long line

File: src/BuildScriptGeneratorCli/Commands/ExecCommand.cs (hunk 12 above)
Comment (dorfire) [reply src/BuildScriptGeneratorCli/Commands/ExecCommand.cs]: `SourceDir` defaults to the current directory, like in `BuildCommand`

File: src/BuildScriptGeneratorCli/Commands/ExecCommand.cs [hunk 14]
```diff
@@ -0,0 +1,109 @@
...
+            }
+
+            int exitCode;
+            using (var timedEvent = logger.LogTimedEvent("ExecCommand"))
+            {
+                var benvCmd = $"{FilePaths.Benv} {StringExtensions.JoinKeyValuePairs(tools)}";
+
+                // Build envelope script
+                var script = new ShellScriptBuilder("\n")
+                    .AddShebang(shellPath)
```
Comment (kichalla) [inline src/BuildScriptGeneratorCli/Commands/ExecCommand.cs]: I think we should add a `set -e` too so that we propagate the right exit code. Add a test for this failure case too.

File: src/BuildScriptGeneratorCli/Commands/ExecCommand.cs [hunk 15]
```diff
@@ -0,0 +1,109 @@
...
+                    .Source(benvCmd)
+                    .AddCommand(Command)
+                    .ToString();
+                logger.LogDebug("Script content:\n{script}", script);
+
+                // Create temporary file to store script
+                var tempScriptPath = Path.GetTempFileName();
+                File.WriteAllText(tempScriptPath, script);
```
Comment (kichalla) [inline src/BuildScriptGeneratorCli/Commands/ExecCommand.cs]: Do we need to set any permissions on the generated script so that it can execute without any problem?

File: src/BuildScriptGeneratorCli/Commands/ExecCommand.cs [hunk 16]
```diff
@@ -0,0 +1,109 @@
...
+                    console.WriteLine($"Temporary script @ {tempScriptPath}:");
+                    console.WriteLine("---");
+                    console.WriteLine(script);
+                    console.WriteLine("---");
+                }
+
+                exitCode = ProcessHelper.RunProcess(
+                    shellPath,
+                    new[] { tempScriptPath },
+                    serviceProvider.GetRequiredService<IOptions<BuildScriptGeneratorOptions>>().Value.SourceDir,
```
Comment (kichalla) [inline src/BuildScriptGeneratorCli/Commands/ExecCommand.cs]: nit: get the options into a variable (after line 36) and use it here instead.

File: tests/Oryx.BuildImage.Tests/ExecCommandTest.cs [hunk 17]
```diff
@@ -0,0 +1,92 @@
...
+                    Assert.Contains(nodeArg, result.StdOut);
+                    Assert.True(result.IsSuccess);
+                    // Actual output from `node --version` starts with a 'v'
+                    Assert.Contains($"v{NodeConstants.NodeLtsVersion}", result.StdOut);
+                },
+                result.GetDebugInfo());
+        }
+
+        [Fact]
+        public void Exec_NodeVersionAndPhpVersion()
```
Comment (kichalla) [inline tests/Oryx.BuildImage.Tests/ExecCommandTest.cs]: Rename the test so that it indicates what the pivot here actually is...CanExecuteMultipleCommands may be?

File: tests/Oryx.BuildImage.Tests/ExecCommandTest.cs [hunk 18]
```diff
@@ -0,0 +1,92 @@
...
+using Microsoft.Oryx.Tests.Common;
+using Xunit;
+using Xunit.Abstractions;
+using Microsoft.Oryx.BuildScriptGenerator.Node;
+using Microsoft.Oryx.Common;
+using Microsoft.Oryx.BuildScriptGenerator.Php;
+
+namespace Microsoft.Oryx.BuildImage.Tests
+{
+    public class ExecCommandTest : SampleAppsTestBase
```
Comment (kichalla) [inline tests/Oryx.BuildImage.Tests/ExecCommandTest.cs]: Add a test where the user provided command exits with a non successful exit code

File: src/BuildScriptGeneratorCli/Commands/ExecCommand.cs [hunk 19]
```diff
@@ -0,0 +1,109 @@
...
+            int exitCode;
+            using (var timedEvent = logger.LogTimedEvent("ExecCommand"))
+            {
+                var benvCmd = $"{FilePaths.Benv} {StringExtensions.JoinKeyValuePairs(tools)}";
+
+                // Build envelope script
+                var script = new ShellScriptBuilder("\n")
+                    .AddShebang(shellPath)
+                    .Source(benvCmd)
+                    .AddCommand(Command)
```
Comment (kichalla) [inline src/BuildScriptGeneratorCli/Commands/ExecCommand.cs]: I think this command should be executed in the context of the source directory, i.e we should probably `cd` into the source directory and then execute the command. This way any user provided relative paths can work fine. Add a test for this scenario too.

File: tests/BuildScriptGeneratorCli.Tests/ExecCommandTest.cs [hunk 20]
```diff
@@ -0,0 +1,57 @@
...
+
+            // Act
+            var exitCode = cmd.OnExecute(new CommandLineApplication(console), console);
+
+            // Assert
+            Assert.Equal(ProcessConstants.ExitSuccess, exitCode);
+        }
+
+        [Fact]
+        public void OnExecute_ShowsErrorAndExits_WhenNoUsableToolsAreDetected()
```
Comment (kichalla) [inline tests/BuildScriptGeneratorCli.Tests/ExecCommandTest.cs]: Since our build image has LTS versions as default, IMO we should not throw this error and should run the provided command without erroring out. Unfortunately I do not know the usage of VS saas team here, but i will give a scenario: Let's say i start out with a directory which is empty and then provide a command like `dotnet new empty`...this should cd into the source directory and run this command and this will create a empty asp.net core project

File: src/BuildScriptGeneratorCli/Commands/ExecCommand.cs (hunk 15 above)
Comment (dorfire) [reply src/BuildScriptGeneratorCli/Commands/ExecCommand.cs]: No - it's not executed, it's just passed as a parameter an read by Bash

File: src/BuildScriptGeneratorCli/Commands/ExecCommand.cs (hunk 16 above)
Comment (dorfire) [reply src/BuildScriptGeneratorCli/Commands/ExecCommand.cs]: that's the purpose of this? I preferred to have one less variable because this isn't used anywhere else

Comment (kichalla) [reply src/BuildScriptGeneratorCli/Commands/ExecCommand.cs]: It just looks ugly IMO. You could argue the same thing for a `using` statement. We do not put the whole namespace at the point we use because we use it in one place.
